pkg/api/response/response.go

```diff
@@ -203,10 +194,13 @@ func Error(status int, message string, err error) *NormalResponse {
 		data["message"] = message
 	}
 
-	var overrideError *OverrideError
-	if errors.As(err, &overrideError) {
-		responseStatus = overrideError.Status
-		data["error"] = overrideError.Error()
+	var errTooManyRequests *plugins.ErrTooManyRequests
```

Comment:
Still feels a bit wrong this package importing "github.com/grafana/grafana/pkg/plugins". I would suggest [metrics.go](https://github.com/grafana/grafana/blob/main/pkg/api/metrics.go) API handlers to handle errors from data source plugins. Something like below. It's not as generic as you've suggested where it's implemented in one place, but still limited to one package/file. WDYT?

```diff
diff --git a/pkg/api/metrics.go b/pkg/api/metrics.go
index d99940ad0d..2d2674a01b 100644
--- a/pkg/api/metrics.go
+++ b/pkg/api/metrics.go
@@ -84,7 +84,7 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext, reqDTO dtos.MetricReq
 
        resp, err := hs.DataService.HandleRequest(c.Req.Context(), ds, request)
        if err != nil {
-               return response.Error(http.StatusInternalServerError, "Metric request error", err)
+               return handleQueryError(err)
        }
 
        // This is insanity... but ¯\_(ツ)_/¯, the current query path looks like:
@@ -97,6 +97,15 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext, reqDTO dtos.MetricReq
        return toMacronResponse(qdr)
 }
 
+func handleQueryError(err error) response.Response {
+       var errTooManyRequests *plugins.ErrTooManyRequests
+       if errors.As(err, &errTooManyRequests) {
+               return response.Error(http.StatusTooManyRequests, errTooManyRequests.Message, err)
+       }
+
+       return response.Error(http.StatusInternalServerError, "Metric request error", err)
+}
+
 func toMacronResponse(qdr *backend.QueryDataResponse) response.Response {
        statusCode := http.StatusOK
        for _, res := range qdr.Responses {
@@ -208,7 +217,7 @@ func (hs *HTTPServer) QueryMetrics(c *models.ReqContext, reqDto dtos.MetricReque
 
        resp, err := hs.DataService.HandleRequest(c.Req.Context(), ds, request)
        if err != nil {
-               return response.Error(http.StatusInternalServerError, "Metric request error", err)
+               return handleQueryError(err)
        }
 
        statusCode := http.StatusOK
```

Reply:
Thanks yeah, I was also thinking about something like this, with the signature of `response.Error()` I wasn't sure if that was ideal place to deal with this. I will refactor it.